### PR TITLE
win/TUI: Use virtual terminal input if available

### LINKED ISSUE #9514: Windows TUI: mouse click to switch window, mouse drag to resize
- `nvim --version`: 0.3.2
- Vim (version:  8.1.22): behaves differently
- Operating system/version: Windows 10 version 1803
- Terminal name/version: [Cmder](https://github.com/cmderdev/cmder/releases), version 1.3.10
- `$TERM`:  In cmder  `settings -> Startup -> Environment `, I use `set TERM=xterm-256color`, `echo &term` inside Neovim shows `builtin_xterm` (for Vim it shows `xterm-256color`).

### Steps to reproduce using `nvim -u NORC`

1.  Use the following minimum setting in `init.vim`:

```
set mouse=a
```

2. open nvim with the above setting:

```
nvim -u init.vim
```

3. Open two files side by side in nvim and use mouse click to try to change panes. try to use mouse drag to resize the window

### Actual behaviour
Can not use mouse click to change windows and can not use mouse drag to resize window in nvim.

But for vim, using mouse click to change windows works (using mouse drag to resize window does not work).

For nvim-qt (using `nvim-qt -- -u init.vim` to open it),  I can switch windows and resize window with mouse correctly.

### Expected behaviour

I can use mouse to change windows and resize window when I use `set mouse=a`.

### PR DESCRIPTION
**Pull request by erw7**
fixes #9514.

The following issues are also fixed on Windows where virtual terminal input is available.

- #8435(may be).
- #11773.

### CODE REVIEW

File: src/nvim/tui/tui.c [hunk 1]
```diff
@@ -1009,17 +1012,51 @@ static void tui_mouse_on(UI *ui)
 {
   TUIData *data = ui->data;
   if (!data->mouse_enabled) {
+#ifdef WIN32
+    const char *term = os_getenv("TERM");
+    bool toggle_uv_vterm_state =
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: what's a more meaningful name for this? maybe `no_vti` ?
```suggestion
    bool no_vti =
```

File: src/nvim/tui/tui.c [hunk 2]
```diff
@@ -1009,17 +1012,51 @@ static void tui_mouse_on(UI *ui)
...
+      term && (terminfo_is_term_family(term, "conemu")
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: is it possible that `term == NULL` but `os_has_vti_working` returns true?

File: src/nvim/os/os_win_console.c [hunk 3]
```diff
@@ -40,3 +40,13 @@ void os_replace_stdout_and_stderr_to_conout(void)
   const int conerr_fd = _open_osfhandle((intptr_t)conout_handle, 0);
   assert(conerr_fd == STDERR_FILENO);
 }
+
+bool os_has_vti_working(void)
```
Comment (justinmk) [inline src/nvim/os/os_win_console.c]:
```suggestion
bool os_has_vti(void)
```

**Comment by justinmk** [comment]
Tested on win10 with cmd.exe, fixes #9514 #11773 (but not #8435) 🚀 

Just some minor questions above, then will merge. Thanks @erw7 !

File: src/nvim/tui/tui.c (hunk 2 above)
Comment (erw7) [reply src/nvim/tui/tui.c]: After merging #11808, I would like to rewrite the condition using `uv_tty_get_vterm_state()`.

File: src/nvim/os/os_win_console.c [hunk 4]
```diff
@@ -40,3 +40,13 @@ void os_replace_stdout_and_stderr_to_conout(void)
   const int conerr_fd = _open_osfhandle((intptr_t)conout_handle, 0);
   assert(conerr_fd == STDERR_FILENO);
 }
+
+bool os_has_vti(void)
+{
+  HANDLE handle = (HANDLE)_get_osfhandle(input_global_fd());
+  DWORD dwMode;
+  if (handle != INVALID_HANDLE_VALUE && GetConsoleMode(handle, &dwMode)) {
+    return dwMode & ENABLE_VIRTUAL_TERMINAL_INPUT;
```
Comment (justinmk) [inline src/nvim/os/os_win_console.c]: this might be a narrowing conversion?

```suggestion
    return !!(dwMode & ENABLE_VIRTUAL_TERMINAL_INPUT);
```

**Comment by justinmk** [comment]
updated to point to https://github.com/neovim/libuv/pull/12

Eager to merge this. Is it still WIP?

**Comment by erw7** [comment]
> Eager to merge this. Is it still WIP?

I want to review the conditions of `use_vti`.

**Comment by erw7** [comment]
I pushed the one that changed the condition.

**Comment by justinmk** [comment]
I pushed a semi-squash (including your latest push d7211f4) .

File: src/nvim/event/stream.c [hunk 5]
```diff
@@ -57,6 +60,11 @@ void stream_init(Loop *loop, Stream *stream, int fd, uv_stream_t *uvstream)
       if (type == UV_TTY) {
         uv_tty_init(&loop->uv, &stream->uv.tty, fd, 0);
         uv_tty_set_mode(&stream->uv.tty, UV_TTY_MODE_RAW);
+        DWORD dwMode;
+        if (GetConsoleMode(stream->uv.tty.handle, &dwMode)) {
+          dwMode |= ENABLE_VIRTUAL_TERMINAL_INPUT;
```
Comment (justinmk) [inline src/nvim/event/stream.c]: This is being done unconditionally?

File: src/nvim/tui/tui.c [hunk 6]
```diff
@@ -1009,17 +1012,51 @@ static void tui_mouse_on(UI *ui)
...
+    bool is_conemu = terminfo_is_term_family(os_getenv("TERM"), "conemu");
+    uv_tty_vtermstate_t state;
+    uv_tty_get_vterm_state(&state);
+    bool is_legacy = state == UV_TTY_UNSUPPORTED;
+    bool use_vti = os_has_vti() && !is_conemu;
+    if (!is_legacy && !use_vti) {
+      flush_buf(ui);
+      uv_tty_set_vterm_state(UV_TTY_UNSUPPORTED);
+    }
+#endif
```
Comment (justinmk) [inline src/nvim/tui/tui.c]: - What is the difference between vterm and VTI? A reference/link would be helpful.
- Why special-case conemu?
- Why temporarily disable vterm before sending mouse on/off requests? Is this working around a problem with our termcodes?


This block (and same code in `tui_mouse_on`) should probably go in `os_win_console.c`, e.g. : 

```
void os_set_wintty(bool enable) {
  static bool is_conemu = ... // Only need to check once, won't to change at runtime.
  if (enable) {
    ...
  } else { 
    ... 
  }
}
```

File: src/nvim/event/stream.c (hunk 5 above)
Comment (erw7) [reply src/nvim/event/stream.c]: Like `ENABLE_VIRTUAL_TERMINAL_PROCESSING`, `ENABLE_VIRTUAL_TERMINAL_INPUT` cannot be determined if it is supported unless it is set with `SetConsoleMode()`. Therefore, set unconditionally.

File: src/nvim/tui/tui.c (hunk 6 above)
Comment (erw7) [reply src/nvim/tui/tui.c]: > - What is the difference between vterm and VTI? A reference/link would be helpful.
> - Why temporarily disable vterm before sending mouse on/off requests? Is this working around a problem with our termcodes?

Comment added. Is the description of the comment clear enough?

> - Why special-case conemu?

I installing and checking ConEmu, there was no need for a special case of ConEmu.

Comment (justinmk) [reply src/nvim/tui/tui.c]: yes, thanks!

**Comment by justinmk** [comment]
appveyor failure:

```
test/functional\ui\mode_spec.lua:73: Row 1 did not match.
Expected:
  |*word(stuff^               |
  |*{0:~                        }|
  |{0:~                        }|
  |{2:-- INSERT --}             |
Actual:
  |*word(stuff)a(stuffa(stuff|
  |*a(stuffa(stuff^           |
  |{0:~                        }|
  |{2:-- INSERT --}             |
```

Restarted appveyor...

**Comment by teto** [comment]
@erw7 I don't have windows but I am willing to merge. Is it ready ?

**Comment by teto** [comment]
also could you rebase to trigger a travis build, for some reason it was not triggered.

**Comment by erw7** [comment]
> I don't have windows but I am willing to merge. Is it ready ?

Yes, I think it's fine to merge. This alleviates the problem(#11061) with some keyboard layouts.

> also could you rebase to trigger a travis build, for some reason it was not triggered.

Done.

**Pull request merged**